File: src/Day12.js
const fs = require("fs")

function sum(obj, blacklistValue) {
    if (typeof obj === "number") {
        return obj
    }
    if (Array.isArray(obj)) {
        return obj.reduce((acc, v) => acc + sum(v, blacklistValue), 0)
    }
    if (typeof obj === "object") {
        const values = Object.values(obj)
        if (blacklistValue != null && values.includes(blacklistValue)) {
            return 0
        }
        return values.reduce((acc, v) => acc + sum(v, blacklistValue), 0)
    }
    return 0
}

function part1() {
    const input = readInput("Day12.txt")
    return sum(JSON.parse(input))
}

function part2() {
    const input = readInput("Day12.txt")
    return sum(JSON.parse(input), "red")
}

function readInput(fileName) {
    return fs.readFileSync(`./${fileName}`).toString()
}

const p1 = part1()
console.log(p1)
const p2 = part2()
console.log(p2);

File: src/Day12.kt
import kotlinx.serialization.json.*

fun main() {

    fun sum(j: JsonElement, blacklist: String?): Int {
        if (j is JsonPrimitive && !j.isString) {
            return j.content.toInt()
        }
        if (j is JsonArray) {
            return j.fold(0) { acc, jsonElement -> acc + sum(jsonElement, blacklist)}
        }
        if (j is JsonObject) {
            if (blacklist != null && j.values.find { it is JsonPrimitive && it.isString && it.content == blacklist } != null) {
                return 0
            }
            return j.values.fold(0) { acc, jsonElement -> acc + sum(jsonElement, blacklist) }
        }
        return 0
    }

    fun part1(input: String): Int {
        val json = Json.parseToJsonElement(input)
        return sum(json, null)
    }

    fun part2(input: String): Int {
        val json = Json.parseToJsonElement(input)
        return sum(json, "red")
    }

    val input = readInput("Day12")[0]
    part1(input).println()
    part2(input).println()
}
